### PR TITLE
Guid Prefix Implementation [13026]

### PR DESCRIPTION
**Pull request by jparisu**

### CODE REVIEW

File: src/cpp/types/endpoint/GuidPrefix.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        ss >> *this;
+    }
+    else
+    {
+        std::stringstream ss(ROS_DISCOVERY_SERVER_GUID_PREFIX_STR_);
+        ss >> *this;
+    }
+
+    // Modify depending the seed
+    // TODO : make available to modify for the whole guid prefix, so is not truncted to 255
```
Comment (juanlofer-eprosima) [inline src/cpp/types/endpoint/GuidPrefix.cpp:55]:
```suggestion
    // TODO : make available to modify for the whole guid prefix, so it is not truncated to 255
```

File: src/cpp/types/endpoint/GuidPrefix.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        this->value[i] = other.value[i];
+    }
+    return *this;
+}
+
+/**
+ * Whether the guid is a valid one
+ *
+ * To be valid, the GuidPrefix and the EntityId must not be invalid / unknown
```
Comment (juanlofer-eprosima) [inline src/cpp/types/endpoint/GuidPrefix.cpp:76]:
```suggestion
 * To be valid, the \c GuidPrefix and the \c EntityId must not be invalid / unknown
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Please fix uncrustify test and remove windows warnings.

File: include/ddsrouter/types/endpoint/GuidPrefix.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include <fastdds/rtps/common/GuidPrefix_t.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+
+/**
+ * @brief GuidPrefix specific class for DDSRouter
+ *
+ * This class references the GuidPrefix_t class in FastDDS, which is a RTPS Participant Unique Id.
+ * This class does not belong to class Guid due to a monumental error in OOP (parallel hierarchy is the future).
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/types/endpoint/GuidPrefix.hpp:31]: :+1:

File: src/cpp/types/endpoint/GuidPrefix.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ */
+
+#include <ddsrouter/types/endpoint/GuidPrefix.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+
+const char* GuidPrefix::SERVER_DEFAULT_GUID_PREFIX_STR_("01.0f.00.00.00.00.00.00.00.00.ca.fe");
+const char* GuidPrefix::ROS_DISCOVERY_SERVER_GUID_PREFIX_STR_("44.53.00.5f.45.50.52.4f.53.49.4d.41");
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/types/endpoint/GuidPrefix.cpp:26]: :+1:

File: src/cpp/types/endpoint/GuidPrefix.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+GuidPrefix::GuidPrefix (bool ros /*= false*/, uint32_t id /*= 0*/) noexcept
+{
+    if (ros)
+    {
+        std::stringstream ss(ROS_DISCOVERY_SERVER_GUID_PREFIX_STR_);
+        ss >> *this;
+    }
+    else
+    {
+        std::stringstream ss(ROS_DISCOVERY_SERVER_GUID_PREFIX_STR_);
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/types/endpoint/GuidPrefix.cpp:50]: :+1:

File: include/ddsrouter/types/endpoint/GuidPrefix.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @param ros : whether use the Discovery Server ROS2 specific guid [Default: false]
+     * @param id : number to seed for the final Guid Prefix [Default: 0]
+     */
+    GuidPrefix (bool ros = false, uint32_t id = 0) noexcept;
+
+    /**
+     * @brief Uses Router default Guid without using ROS Discovery Server
+     *
+     * @param id number to seed for the final GUID [Default: 0]
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/types/endpoint/GuidPrefix.hpp:73]: :+1:

File: include/ddsrouter/types/endpoint/GuidPrefix.hpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * Use \c id as seed to create an arbitrary GuidPrefix.
+     * It could be used as default Guid the ROS2 Discovery Server Guid [44.53.00.5f.45.50.52.4f.53.49.4d.41]
+     * or the DDSRouter Discovery Server Guid [01.0f.00.00.00.00.00.00.00.00.ca.fe].
+     *
+     * The value in 3rd position ([2]) is the one set by seed.
+     *
+     * @todo use the seed of \c id to modify the whole guid and not only one of the 12 values.
+     *
+     * @param ros : whether use the Discovery Server ROS2 specific guid [Default: false]
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/types/endpoint/GuidPrefix.hpp:65]: :+1:

File: src/cpp/types/endpoint/GuidPrefix.cpp (hunk 2 above)
Comment (jparisu) [reply src/cpp/types/endpoint/GuidPrefix.cpp:76]: :+1:

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**